### PR TITLE
Correct calculation of `AbiType` sizes

### LINKED ISSUE #538: Incorrectly calculating size of struct witness lengths.
# Description

## Aim

I want to be able to calculate the size of a struct in terms of the number of field elements which are written into the witness map.

## Expected behavior

The struct's size should be the sum of the sizes of its fields.

## Bug

Here in `process_abi_with_input` we report that a struct uses a number of field elements equal to the number of fields (properties) it has:
https://github.com/noir-lang/noir/blob/ac131bd3526cb6ba40c461bf3efd6c21acee0bf1/crates/nargo/src/cli/prove_cmd.rs#L57

We're then making the assumption that each field corresponds to a single field element. This works for structs which are made up from `Field`/`Integer` types only but not for anything more complex.

https://github.com/noir-lang/noir/blob/ac131bd3526cb6ba40c461bf3efd6c21acee0bf1/crates/nargo/tests/test_data/struct_inputs/src/foo/bar.nr#L3-L6

`barStruct` is represented by 3 field elements however `process_abi_with_input` returns that if we were to read it out of the witness map then it should only contain 2 field elements.

https://github.com/noir-lang/noir/blob/ac131bd3526cb6ba40c461bf3efd6c21acee0bf1/crates/nargo/tests/test_data/struct_inputs/src/foo.nr#L3-L5

`fooStruct` is reported as being a single field element despite containing a value which is made of 2 field elements.

# To reproduce

Not reproducible as setting a struct as a return value results in a `no support for == with types struct/tuple and struct/tuple` error.

# Additional context

(If applicable.)

### PR DESCRIPTION
**Pull request by TomAFrench**
# Related issue(s)

fixes #538 

# Description

## Summary of changes

I've fixed the calculation of `return_witness_len` in `process_abi_with_input` to properly calculate the correct length for arbitrary arrays or structs.

I've done this by adding a `size` method on `AbiType` to calculate the number of field elements required for a type by recursively calculating the size of any subtypes.

This may overlap with the intended usage of `num_elements` but I wasn't 100% sure if that was referring to "elements" as in field elements or as component subtypes. `num_elements` is unused however so we could just nuke it.

## Dependency additions / changes

N/A

## Test additions / changes

N/A

# Checklist

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [x] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [x] I have reviewed the changes on GitHub, line by line.
- [x] I have ensured all changes are covered in the description.

# Additional context

While we currently can't fall afoul of this issue currently. Fixing this allows us to generalise decoding all the circuit inputs so we won't need to pass around the return value witness index in order to support passing an undefined return value to the prover.

### CODE REVIEW

**Review by jfecher — CHANGES_REQUESTED** [review]
This looks good to me, lets just rename `size` to something a bit clearer like - for example - `field_count`.

File: crates/noirc_abi/src/lib.rs [hunk 1]
```diff
@@ -66,6 +66,17 @@ impl AbiType {
         }
     }
 
+    /// Returns the number of field elements required to represent the type once encoded.
+    pub fn field_count(&self) -> usize {
+        match self {
+            AbiType::Field(_) | AbiType::Integer { .. } => 1,
+            AbiType::Array { visibility: _, length, typ } => typ.field_count() * (*length as usize),
```
Comment (jfecher) [inline crates/noirc_abi/src/lib.rs:73]: Apologies for not catching this earlier - if we only need `field_count` in one place expecting a `u32` we might as well return a `u32` directly. I'd rather convert to `u32` directly than having to convert to usize here then to u32 later.

Comment (TomAFrench) [reply crates/noirc_abi/src/lib.rs:73]: No worries. I wanted to keep `usize` for consistency with `AbiType.num_elements` but happy to change.

**Review by jfecher — APPROVED** [review]
Thank you!

**Review by vezenovm — APPROVED** [review]
Looks good.

**Pull request merged**